primitives/block-builder/src/lib.rs

```diff
@@ -20,18 +20,24 @@
 #![cfg_attr(not(feature = "std"), no_std)]
 
 use sp_inherents::{CheckInherentsResult, InherentData};
-use sp_runtime::{traits::Block as BlockT, ApplyExtrinsicResult};
+use sp_runtime::{
+	legacy::byte_sized_error::ApplyExtrinsicResult as ApplyExtrinsicResultBeforeV6,
+	traits::Block as BlockT, ApplyExtrinsicResult,
+};
 
 sp_api::decl_runtime_apis! {
 	/// The `BlockBuilder` api trait that provides the required functionality for building a block.
-	#[api_version(5)]
+	#[api_version(6)]
```

Comment:
@gavofyork Here we've bumped the API version of `BlockBuilder`, which contains the `apply_extrinsic` function and I believe it is what the client calls which returns the modified `Result`/`Error`. We've made it so that we always check for the API version first before we attempt to submit an extrinsic, which guarantees that we use the correct `Result` for decoding.

The one lingering problem that I can think of -- if I'm not mistaken that `applly_extrinsic` is really the only place where clients can obtain a `DispatchError` -- is that some pallets may have put a `DispatchError` in storage for whatever reason, in which case, they'll have to perform a runtime migration in order to decode the stored error value.

Reply:
But this surely only helps with migrating to the changes in this PR. It doesn't seem to help e.g. when a runtime is upgraded and an extrinsic returns an error which cannot be decoded by the client.

Reply:
Ok never mind. I see that the client doesn't seem to try to decode the error anyway.

Reply:
We make static guarantees on the maximum size of the error type. If the `MaxEncodedLen` of a pallet error exceeds 4 bytes, then the pallet simply would not compile.

Reply:
> Ok never mind. I see that the client doesn't seem to try to decode the error anyway.

I think it does actually, when the client is calling `apply_extrinsics`, and we have the version-aware logic codec in babebfc86de9d4dd9906ea9f87af178647c82d8c.